### PR TITLE
python3Packages.tidalapi: fix build error; python3Packages.mpegdash: init at 0.4.0

### PR DESCRIPTION
**Pull request by drawbu**
ZHF: #309482

## Description of changes

fixed tidalapi by adding missing dependecies
mpegdash was a requirement for tidalapi, so I figured we should add it to nixpkgs

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by drawbu** [comment]
Oops. Didn't saw you were not a maintainer anymore. Sorry :/

**Comment by drawbu** [comment]
Result of `nixpkgs-review pr 310439` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>mopidy-tidal</li>
    <li>mopidy-tidal.dist</li>
    <li>python311Packages.mpegdash</li>
    <li>python311Packages.mpegdash.dist</li>
    <li>python311Packages.tidalapi</li>
    <li>python311Packages.tidalapi.dist</li>
    <li>python312Packages.mpegdash</li>
    <li>python312Packages.mpegdash.dist</li>
    <li>python312Packages.tidalapi</li>
    <li>python312Packages.tidalapi.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/mpegdash/default.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  unittestCheckHook,
```
Comment (annaleeleaves) [inline pkgs/development/python-modules/mpegdash/default.nix:5]:
```suggestion
  pytestCheckHook,
```

File: pkgs/development/python-modules/mpegdash/default.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  };
+
+  buildInputs = [ unittestCheckHook ];
+
+  # I took the liberty to remove the test case that requires network access
+  # I think is better to remove a single test than all
+  preCheck = ''
+    substituteInPlace tests/test_xmltompd.py \
+      --replace-warn "def test_xml2mpd_from_url(self):" "def _test_xml2mpd_from_url(self):"
+  '';
```
Comment (annaleeleaves) [inline pkgs/development/python-modules/mpegdash/default.nix:25]:
```suggestion
  buildInputs = [ pytestCheckHook ];

  disabledTests = [
    # requires network access
    "test_xml2mpd_from_url"
  ];
```

**Review by ghost — COMMENTED** [review]
looks good though the commits need to be fixed up and disabling tests can be fixed (suggestion inline)

- mpegdash commits can be squashed and then re-ordered so they come before the tidalapi commit and only contain the changes that add mpegdash package.
- tidalapi commits come next and just contain changes to tidalapi package
- commit messages for python need to specify a python version to trigger ofborg automated builds. eg:

  `python311Packages.tidalapi: fix build failure`
  `python311Packages.mpegdash: init at 0.4.0`
  or to trigger python 3.11 and 3.12
  ` python31{1,2}Packages.mpegdash: init at 0.4.0`
  but most people just trigger one.
  https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building

File: pkgs/development/python-modules/mpegdash/default.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  unittestCheckHook,
+}:
+buildPythonPackage rec {
+  pname = "mpegdash";
+  version = "0.4.0";
```
Comment (annaleeleaves) [inline pkgs/development/python-modules/mpegdash/default.nix:9]:
```suggestion
  version = "0.4.0";
  pyproject = true;
```

File: pkgs/development/python-modules/tidalapi/default.nix [hunk 4]
```diff
@@ -17,26 +20,24 @@ buildPythonPackage rec {
...
   propagatedBuildInputs = [
     requests
     python-dateutil
+    mpegdash
+    isodate
+    ratelimit
+    typing-extensions
   ];
 
-  doCheck = false; # tests require internet access
```
Comment (annaleeleaves) [inline pkgs/development/python-modules/tidalapi/default.nix:29]: the tests in the github source all require internet access, so this line should not be removed.

**Review by ghost — COMMENTED** [review]
diffs look good but commits still could use a minor adjustment as there are changes to tidalapi in the mpegdash commit.

<details><summary>example git commands</summary>

the commands below create `test-branch` from `fix-tidal` and then switches to the `test-branch` and performs some git transformations to undo the last three commits without modifying the files. then re-adding and committing the files. you can verify that this does what you want and then switch to the `fix-tidal` branch, perform the same commands but skip  `git switch -c test-branch fix-tidal`, and then force-push the change.

```console
git switch -c test-branch fix-tidal
git reset HEAD~3
git add pkgs/development/python-modules/mpegdash/ pkgs/top-level/python-packages.nix
git commit -m "python31{1,2}Packages.mpegdash: init at 0.4.0"
git commit -m "python31{1,2}Packages.tidalapi: fix build failure" pkgs/development/python-modules/tidalapi/default.nix

```
</details>

**Review by ghost — APPROVED** [review]
LGTM

Result of `nixpkgs-review pr 310439` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>mopidy-tidal</li>
    <li>mopidy-tidal.dist</li>
    <li>python311Packages.mpegdash</li>
    <li>python311Packages.mpegdash.dist</li>
    <li>python311Packages.tidalapi</li>
    <li>python311Packages.tidalapi.dist</li>
    <li>python312Packages.mpegdash</li>
    <li>python312Packages.mpegdash.dist</li>
    <li>python312Packages.tidalapi</li>
    <li>python312Packages.tidalapi.dist</li>
  </ul>
</details>

**Pull request merged**